### DISCUSSION TITLE
Amplitude calibration

### DISCUSSION THREAD

**Discussion by allico** [Cochlear-Research/nucleus-toolbox#4] [Q&A] [answered]
First of all, thank you so much for the newly released version of this toolbox. I've been reviewing the source code to learn how to use it and better understand how to translate our current project, based on NMT 4.31, to this newer version.

I'm particularly interested in the implementation of ACE and I've been primarily reviewing the code `ACE_map.m`. Here are my questions/comments that I'm hoping to get some clarifications for.

- `FFT_filterbank_proc.m`: I believe that the amplitude after computing the FFT is missing the conversion factor of going from two-sided FFT to one-sided FFT.
- `Ensure_amplitude_params.m`: what is the purpose of computing `p.gain_dB` here? I believe that, as is, it's over-amplifying the input signal. For example, for a speech signal at C-SPL, its `rms = 0.0224` and the peak would be at `0.0793` (based on default parameters). Using the default gain, `p.gain_dB = 36`, we would produce a speech signal with peak amplitude at `5.0059`. This signal would get clipped by `LGF_proc.m` that has a saturation level of `1.0`.
- What is the difference between `FE_AGC_proc.m` and `Freedom_AGC_proc.m`? Are they meant to simulate the AGC stage of different processors? Which one should be preferred?

Based on the testing I've done so far, my speech signals generate an output after the `LGF_proc.m` process that is generally clipped during the voiced segments. Most of my signals have an amplitude less than 0.5, and are being amplified to the default target of 65 dB SPL. I hope you can provide some insight as for why the signals are being amplified so much.

Thanks.

**Comment by BrettSwansonPhD** [comment, marked as answer]
# FFT filterbank
 
With default parameters (128 point FFT, Hann window), the gain of each channel of FFT_filterbank_proc is 32. This means that a sinusoid with peak amplitude 1.0 at the centre frequency of an FFT bin gives an output with peak amplitude 32.0.  Perhaps it would have been better to normalize the gain to unity. However, FFT_filterbank_proc is not used directly in ACE_map; the filterbank is implemented by FFT_VS_filterbank_proc, which calls FFT_filterbank_proc then uses a vector-sum to combine the 65 FFT channels into the desired number of bands (default 22). By default, p.equalise = 1 and FFT_VS_filterbank_proc normalizes the gain of each band to unity, so the gain of FFT_filterbank_proc doesn’t really matter.
 
# Amplitude calibration
 
The default parameters of each proc were chosen so that when a proc is used by itself, typical signals would be in the range 0 to 1, which is convenient for testing. However, when the entire signal path is constructed in ACE_map, the parameters of the individual procs need to be co-ordinated. This is the purpose of Ensure_amplitude_params.
Firstly, p.reference_dB_SPL (default 95 dB) determines the sound pressure level that is represented by a full-scale sinusoid (peak value +1, RMS value -3 dB). This is used in Audio_proc to scale the values read from a wav file so that they represent the sound pressure level given in p.audio_dB_SPL (default 65 dB SPL).
As an example:
 
    >> p = ACE_map;
    >> c = Process_chain(p, 'L001s04r.wav');
 
The output of Audio_proc is in c{1}; as you mentioned, it has been scaled to represent 65 dB SPL, i.e. its RMS value is 30 dB below the RMS value of a full scale sinusoid:
 
    >> To_dB(rms(c{1}))
    ans =
      -33.0103
  
The microphone frequency response attenuates low frequencies and reduces the RMS; but the AGC is mostly inactive for this signal. The AGC output is in c{3}, and its peak value is:
 
    >> max(abs(c{3}))
    ans =
        0.1110
 
Each filter in FFT_VS_filterbank_proc has unity gain for a sinusoid at the centre frequency of the band. However, for a speech signal, the relationship between the peak level at the input to the filterbank and the output level of each band envelope is not straightforward. The peak value of the frequency-time matrix (FTM) across all time samples of all bands is:
 
    >> max(c{5}(:))
    ans =
        0.0634
 
However, the vast majority of the FTM samples are substantially below this peak.
Our definition of C-SPL is that a speech signal at C-SPL should occasionally reach C-level on some channels.
This means that some channels should occasionally reach p.sat_level (default 1.0) at the input to LGF_proc.
Ensure_amplitude_params calculates p.gain_dB using a rule-of-thumb that the speech signal should have an RMS value 6 dB lower than the pure tone that would just reach p.sat_level. Hence if we start with C_SPL 30 dB below the reference level, then we need 30 + 6 = 36 dB of gain after the filterbank.
 
The Channel-Magnitude sequence is in c{9}:
 
    >> c{9}
    ans =
      struct with fields:
 
          channels: [44856×1 double]
        magnitudes: [44856×1 double]
        periods_us: 85.4000
 
Out of the total 44856 pulses, the number of pulses that have magnitude of 1 (and hence will get mapped to the upper current level i.e. C-level) is 496, i.e. about 1% of pulses reach C-level:
 
    >> msat = (c{9}.magnitudes == 1);
    >> sum(msat)
    ans =
       496
    >> sum(msat)/length(msat)
    ans =
        0.0111
 
Plot_sequence shows that there is a good range of current levels across most channels.
 
# Front-end AGC
 
FE_AGC_proc and Freedom_AGC_proc are both front-end automatic gain control. Freedom_AGC_proc is the algorithm that was used in the Freedom processor. FE_AGC_proc is a simplified algorithm, written so that the principle of operation is easier to understand.
The processors after Freedom have a different AGC which is not available in Nucleus Toolbox.

**Reply by allico** [reply]
Thank you so much for your responses. After reviewing your answers and doing some testing myself, I have the following follow-up questions.

# FFT_filterbank_proc

I agree that the filterbank is normalized so that the maximum gain of each channel is equal to 1. However, it doesn't take into account the output of the FFT computed in FFT_filterbank_proc. For example, I modify the processing in FFT_filterbank_proc to include the factor of 2 for going from two-sided to one-sided, and the output of the filterbank also increased by a factor of 2, leading me to conclude that this factor does has an effect.

# Amplitude calibration

I recreated the examples you included in your response, and I believe that computing the number of saturated samples over the total number of samples is masking what happens at each channel. For instance, if we compute the percentage of samples that are at saturation level at the output of LGF_proc (c{8}) for each channel, we would obtain the following percentages:

```
>> chSat = (c{8} == 1);
>> sum(chSat, 2)
ans =
   197
   66
   67
   34
   0
   0
   ...
>> sum(chSat, 2) / length(chSat, 2) * 100
ans =
   5.2702
   1.7657
   1.7924
   0.9096
   0
   0
   ...
```

I truncated the output to save space, but most remaining channels have very few instances of saturation. I think that makes sense as I wouldn't expect a lot of energy in those channels. However, out of the 496 samples that were saturated, about half of them are concentrated in channels 1 and 2 (electrodes 22 and 21, respectively). Moreover, the numbers shown above increase when I switch the directivity of the microphone from 'dir' (default) to 'omni', with the total number of saturated samples increasing to 2343, and we would obtain the following percentages per channel:

```
>> chSat = (c{8} == 1);
>> sum(chSat, 2)
ans =
   1264
   563
   377
   74
   3
   0
   ...
>> sum(chSat, 2) / length(chSat, 2) * 100
ans =
   33.8149
   15.0615
   10.0856
   1.9797
   0.0803
   0
   ...
```

As you can see above, channel 1 is saturated about a third of the time, with channels 2 and 3 saturated at least 10% each.

Please correct me if I'm wrong, but my understanding is that the Nucleus 6 (processor provided to us) uses a dual omni mic array, where the Standard Directionality setting corresponds to an omni pattern, and this setting is preferred for quiet speech environments. Using this information, I would be inclined to select omni over the default, unless there are other reasons that I'm not aware of for using the default directivity.

For the study we are working on, we are measuring baseline performance using the Nucleus 6 processor and speech presented in free field. Consequently, it is important to us to properly model the mic used by our processor to minimize sources of errors.

# Sound quality omni vs dir

When listening to the output of the module Freedom_micrpphone_proc (c{2}) using both directivity options, I personally believe that 'omni' sounds better than 'dir'. However, the low frequencies that are not attenuated by the flatter frequency response of omni (when compared to dir that resembles a pass-band frequency response) result in a significant number of samples mapped to saturation levels, as shown in the percentages described above.

Do the directivity settings model the behavior of a Nucleus 6 processor? And is it accurate to say that the omni setting is modeling the behavior of the Standard Directionality setting?

# Speech vs pure tone offset

Unless I misunderstood, I think there is a discrepancy between your response and the comments in the code of Ensure_amplitude_params. You said that the rule-of-thumb is that a speech signal should have an RMS value 6 dB lower than the pure tone. However, the comments in the code say that saturation is achieved with a tone that is 6 dB lower than the speech signal. The calculated gain matches your answer posted here, not the comments in the code.

Moreover, if a pure tone and speech signals achieve the same dB SPL level for different RMS values, why is this offset not included in the module `Audio_proc` when calculating the gain by comparing the input audio to the reference dB SPL, which was measured using a pure tone?
